### DISCUSSION TITLE
autoresearch-robotics: adapted for robotics with mujoco simulation feedback

### DISCUSSION THREAD

**Discussion by jellyheadandrew** [karpathy/autoresearch#60] [Show and tell]
**Adapted autoresearch for robotics**

Fork: [https://github.com/jellyheadandrew/autoresearch-robotics](https://github.com/jellyheadandrew/autoresearch-robotics)

Same core loop: agent modifies the training code, runs the experiment, checks if the result improved, keeps or discards, and repeats autonomously.

The key adaptation is replacing the LLM training loop with a robotics simulation feedback loop - the agent optimizes policy training code against task success rate AND renderings from MuJoCo, instead of validation loss.

![teaser-gif](https://github.com/user-attachments/assets/a6b5564b-9b90-49b4-b45a-306245339193)

**What's different**

- Visual feedback. After each experiment, MuJoCo renders the robot's behavior and Claude Vision analyzes the frames. The agent sees what the robot is doing wrong, not just numbers 

**Experimentally, I feel it provides better qualitative feedbacks for next trial.**

(Example1)
> GRASPS cube! but cant transport to goal (dist 0.22)	discard	balanced throughput+reward shaping (58K steps, 11K updates)

(Example2)
> inconsistent gripper orientation, no contact	discard	vectorized HER + N_UPDATES=10 (55K steps but too few updates)

I ran experiments on simple robot-learning task (FetchReach). The agent started from an SAC+HER baseline and autonomously discovered that a simple proportional controller solves the task. I'm currently running more complex tasks (FetchPush and FetchPickPlace), and will try VLAs after I get some GPU compute credits.

<img width="1482" height="878" alt="results_plot_fetchreach" src="https://github.com/user-attachments/assets/53f55980-2c6e-41b3-9bb5-12ee1366c864" />

Tested on an RTX 2080.

Would love feedback from anyone working on robotics or sim-to-real! I'd be very grateful if anyone finds it useful.

**Comment by aniruddhaadak80** [comment]
This is a brilliant application of the framework. Swapping the language model training out for a mujoco robotics simulation is such a creative use case.

I am curious how the agent handles the continuous state space feedback compared to discrete token loss. Really great work on this.

**Reply by jellyheadandrew** [reply]
Thank you!

The keep/discard gate is still scalar, task success rate averaged over eval rollouts. So structurally it's the same ratchet as the original. Joint angles, end-effector positions, contact forces etc. all get collapsed into that one number at eval time.

Where it diverges is between experiments. In the original, the agent only knows "val_bpb went up/down" and has to guess why. For here I believe Claude Vision looks at MuJoCo renderings and can reason about what's physically going wrong. "gripper oriented wrong," "made contact but couldn't lift." The qualitative signal drives the next hypothesis, not the keep/discard decision.

The honest challenge vs the LLM setting is variance. RL success rates are way noisier than validation loss, so you need enough eval rollouts to trust the gate. Still iterating on that, especially for harder tasks like FetchPickAndPlace.

**Comment by Nyrok** [comment]
The separation you describe is the interesting part: task success rate serves as the keep/discard gate, but the visual qualitative reasoning (what the robot did wrong) drives the next hypothesis. In vanilla autoresearch both functions collapse into a single scalar, val_bpb goes up or down and the agent has to infer direction from that alone.

Separating them forces the agent to articulate why before it iterates. The qualitative signal isn't the gate, it's the hypothesis generator. That's a cleaner design, and it likely explains why you're getting better directional feedback even with noisy success rates.

It also maps directly to how you'd want to structure the program.md for robotics: goal (maximize task success rate), constraints (stop criteria), and chain of thought (reason over visual frames to form next hypothesis) as distinct labeled sections rather than one prose block.

I've been building [flompt](https://flompt.dev) for exactly that, a visual builder that decomposes task instructions into 12 typed semantic blocks and compiles to Claude-optimized XML. Open-source: github.com/Nyrok/flompt
